### PR TITLE
Run scripts in threads

### PR DESCRIPTION
**Pull request by chamini2**

### CODE REVIEW

File: src/fal/cli/fal_runner.py [hunk 1]
```diff
@@ -46,9 +46,7 @@ def fal_run(
         )
 
 
-def _create_fal_dbt(
-    args: argparse.Namespace,
-):
+def _create_fal_dbt(args: argparse.Namespace, args_dict: Dict[str, Any]):
```
Comment (mkam039) [inline src/fal/cli/fal_runner.py:49]: `args` and `args_dict` both contain args, right? Can we just use one?

File: src/fal/run_scripts.py [hunk 2]
```diff
@@ -46,48 +50,98 @@ class ContextConfig:
...
+                    logger.debug("Finished GLOBAL script {}", script.path)
+                else:
+                    logger.debug(
+                        "Finished script {} for model {}",
+                        script.path,
+                        script.model.name,
+                    )
+                done_scripts[context.unique_id] = True
+                if err is not None:
+                    # TODO: what to do?
```
Comment (mkam039) [inline src/fal/run_scripts.py:93]: Once all the scripts are done, we should call `sys.exit` with non-zero exit code.

https://docs.python.org/3.8/library/sys.html#sys.exit

File: src/fal/run_scripts.py [hunk 3]
```diff
@@ -46,48 +50,98 @@ class ContextConfig:
...
+            if _is_global_script(script):
+                logger.debug("Running GLOBAL script {}", script.path)
+            else:
+                logger.debug(
+                    "Running script {} for model {}",
+                    script.path,
+                    script.model.name,
+                )
+
+            def callback(tpl: Tuple[FalScript, Context, Exception]):
```
Comment (mkam039) [inline src/fal/run_scripts.py:81]: Clearer function and argument names could be helpful.

**Review by mkam039 — COMMENTED** [review]
Looks very nice! Integration tests seem to be failing for some cases though.

File: src/fal/run_scripts.py (hunk 2 above)
Comment (chamini2) [reply src/fal/run_scripts.py:93]: this case is for when a script had an error

File: src/fal/cli/fal_runner.py (hunk 1 above)
Comment (chamini2) [reply src/fal/cli/fal_runner.py:49]: I guess we could. Does not feel part of this PR though. Can we do it in another one?

**Review by mkam039 — APPROVED** [review]
Great!

File: src/fal/fal_script.py [hunk 4]
```diff
@@ -26,9 +26,10 @@ def exec(self, context, faldbt: FalDbt):
         sys.path.append(local_path)
 
         with open(self.path) as file:
-            a_script = file.read()
+            script_ast = compile(file.read(), file.name, "exec", dont_inherit=True)
```
Comment (turbo1912) [inline src/fal/fal_script.py:29]: why do we need to do this?

File: src/fal/fal_script.py [hunk 5]
```diff
@@ -40,5 +41,7 @@ def exec(self, context, faldbt: FalDbt):
                     "list_sources": faldbt.list_sources,
                     "list_features": faldbt.list_features,
                 },
+                locals,
             )
             sys.path.remove(local_path)
+        return locals.get("RESULT")
```
Comment (turbo1912) [inline src/fal/fal_script.py:47]: change the return type of this function

Comment (turbo1912) [inline src/fal/fal_script.py:47]: I dont understand why we are doing this locals thing?

File: src/fal/run_scripts.py [hunk 6]
```diff
@@ -48,46 +52,92 @@ class Context:
...
-        context_config = ContextConfig(_get_target_path(faldbt._config))
-        context = Context(current_model=None, config=context_config)
+        except KeyboardInterrupt:
+            pool.close()
+            pool.terminate()
+            pool.join()
+            raise
+
+    logger.debug("Script results: {}", results)
+    return results
```
Comment (turbo1912) [inline src/fal/run_scripts.py:109]: why did we decide to return results from scripts?

File: src/fal/run_scripts.py [hunk 7]
```diff
@@ -2,18 +2,22 @@
 import os
 from typing import Dict, Any, List, Union
 from dataclasses import dataclass
-from dbt.config.runtime import RuntimeConfig
 from pathlib import Path
 
-from dbt.contracts.results import NodeStatus
+from multiprocessing.dummy import Pool as ThreadPool
```
Comment (turbo1912) [inline src/fal/run_scripts.py:7]: is this right? 

"multiprocessing.dummy replicates the API of multiprocessing but is no more than a wrapper around the threading module."

Comment (chamini2) [reply src/fal/run_scripts.py:7]: Yes, it imitates the `multiprocessing` API, which is dedicated to forks but uses threads instead.

File: src/fal/fal_script.py (hunk 4 above)
Comment (chamini2) [reply src/fal/fal_script.py:29]: I think this and all the results stuff can actually go in a different PR. I will do that.

**Review by turbo1912 — APPROVED** [review]

**Pull request merged**